src/basic/terminal-util.c

```diff
@@ -1206,38 +1208,65 @@ bool terminal_is_dumb(void) {
         return getenv_terminal_is_dumb();
 }
 
-bool colors_enabled(void) {
-
-        /* Returns true if colors are considered supported on our stdout. For that we check $SYSTEMD_COLORS first
-         * (which is the explicit way to turn colors on/off). If that didn't work we turn colors off unless we are on a
-         * TTY. And if we are on a TTY we turn it off if $TERM is set to "dumb". There's one special tweak though: if
-         * we are PID 1 then we do not check whether we are connected to a TTY, because we don't keep /dev/console open
-         * continuously due to fear of SAK, and hence things are a bit weird. */
-
-        if (cached_colors_enabled < 0) {
-                int val;
+static ColorMode parse_systemd_colors(void) {
+        const char *e;
+        int r;
 
-                val = getenv_bool("SYSTEMD_COLORS");
-                if (val >= 0)
-                        cached_colors_enabled = val;
+        e = getenv("SYSTEMD_COLORS");
+        if (!e)
+                return _COLOR_INVALID;
+        if (streq(e, "16"))
+                return COLOR_16;
+        if (streq(e, "256"))
+                return COLOR_256;
+        r = parse_boolean(e);
+        if (r >= 0)
+                return r > 0 ? COLOR_ON : COLOR_OFF;
+        return _COLOR_INVALID;
+}
 
+ColorMode get_color_mode(void) {
+
+        /* Returns the mode used to choose output colors. The possible modes are COLOR_OFF for no colors,
+         * COLOR_16 for only the base 16 ANSI colors, COLOR_256 for more colors and COLOR_ON for unrestricted
+         * color output. For that we check $SYSTEMD_COLORS first (which is the explicit way to
+         * change the mode). If that didn't work we turn colors off unless we are on a TTY. And if we are on a TTY
+         * we turn it off if $TERM is set to "dumb". There's one special tweak though: if we are PID 1 then we do not
+         * check whether we are connected to a TTY, because we don't keep /dev/console open continuously due to fear
+         * of SAK, and hence things are a bit weird. */
+        ColorMode m;
+
+        if (cached_color_mode < 0) {
+                m = parse_systemd_colors();
+                if (m >= 0)
+                        cached_color_mode = m;
                 else if (getenv("NO_COLOR"))
                         /* We only check for the presence of the variable; value is ignored. */
-                        cached_colors_enabled = false;
+                        cached_color_mode = COLOR_OFF;
 
                 else if (getpid_cached() == 1)
                         /* PID1 outputs to the console without holding it open all the time */
-                        cached_colors_enabled = !getenv_terminal_is_dumb();
+                        cached_color_mode = getenv_terminal_is_dumb() ? COLOR_OFF : COLOR_16;
```

Comment:
this deserves some explanation in the comment, i.e. why 16 colors here instead of the full thing.

Reply:
If I understand this right, when systemd is running as PID1 it outputs to the Linux console, which doesn't support more than 16 colors. Is this correct?

Reply:
Seems so:
![Screenshot from 2020-12-11 12-03-18](https://user-images.githubusercontent.com/349618/101896202-e7891f80-3ba8-11eb-82ee-4596a54b2a4d.png)
![Screenshot from 2020-12-11 12-04-05](https://user-images.githubusercontent.com/349618/101896272-01c2fd80-3ba9-11eb-93f2-4a1ef73ae5e5.png)
<details>

```bash
curl -s https://gist.githubusercontent.com/HaleTom/89ffe32783f89f403bba96bd7bcd1263/raw/ | bash
```
</details>

Reply:
Ok, I put a note in the comment above.

Reply:
hmm, so i boot up systemd daily in an nspawn container, where the output goes to my gnome terminal... Given that the Linux kernel console actually handles 256 colors OKish these days (by mapping them back to the closest color in the 16 color palette), I think we should do 256 colors on the console too, no?